collect_app/src/main/java/org/odk/collect/android/widgets/BaseArbitraryFileWidget.java

```diff
@@ -0,0 +1,76 @@
+package org.odk.collect.android.widgets;
+
+import android.content.Context;
+
+import androidx.annotation.NonNull;
+
+import org.javarosa.core.model.data.IAnswerData;
+import org.javarosa.core.model.data.StringData;
+import org.odk.collect.android.formentry.questions.QuestionDetails;
+import org.odk.collect.android.utilities.MediaUtils;
+import org.odk.collect.android.utilities.QuestionMediaManager;
+import org.odk.collect.android.widgets.interfaces.FileWidget;
+import org.odk.collect.android.widgets.interfaces.WidgetDataReceiver;
+import org.odk.collect.android.widgets.utilities.WaitingForDataRegistry;
+
+import java.io.File;
+
+import timber.log.Timber;
+
+public abstract class BaseArbitraryFileWidget extends QuestionWidget implements FileWidget, WidgetDataReceiver  {
+    @NonNull
+    protected final MediaUtils mediaUtils;
+
+    private final QuestionMediaManager questionMediaManager;
+    protected final WaitingForDataRegistry waitingForDataRegistry;
+
+    protected File answerFile;
+
+    public BaseArbitraryFileWidget(Context context, QuestionDetails questionDetails, @NonNull MediaUtils mediaUtils,
+                                   QuestionMediaManager questionMediaManager, WaitingForDataRegistry waitingForDataRegistry) {
+        super(context, questionDetails);
+        this.mediaUtils = mediaUtils;
+        this.questionMediaManager = questionMediaManager;
+        this.waitingForDataRegistry = waitingForDataRegistry;
+    }
+
+    @Override
+    public IAnswerData getAnswer() {
+        return answerFile != null ? new StringData(answerFile.getName()) : null;
+    }
+
+    @Override
+    public void deleteFile() {
+        questionMediaManager.deleteAnswerFile(getFormEntryPrompt().getIndex().toString(), answerFile.getAbsolutePath());
+        answerFile = null;
+    }
+
+    @Override
+    public void setData(Object object) {
+        if (answerFile != null) {
+            deleteFile();
+        }
+
+        if (object instanceof File) {
+            answerFile = (File) object;
+            if (answerFile.exists()) {
+                questionMediaManager.replaceAnswerFile(getFormEntryPrompt().getIndex().toString(), answerFile.getAbsolutePath());
+                showAnswerText();
+                widgetValueChanged();
+            } else {
+                answerFile = null;
+                Timber.e("Inserting Arbitrary file FAILED");
+            }
+        } else {
+            Timber.e("FileWidget's setBinaryData must receive a File or Uri object.");
```

Comment:
Currently `Uri` objects aren't supported so this is misleading. In order for it to be actionable, we should log the actual file type received.

Reply:
Good catch! I removed that `or Uri object`. However i'm not sure what you mean by:
> we should log the actual file type received

This else block is a result of `!(object instanceof File)` so what file types you do you want to log here? Or maybe you want to add such a log above where it's really a file?

Reply:
I meant we could log `object.getClass()` so at least then we'd be able to track down where a bad call may have come from. In practice it feels like we can control the call site well enough to be confident everything coming in will be a `File` so hopefully this branch would never be hit.

Reply:
Ok I see. I changed it and now I log object.getClass() if the object is not null because if it is null it most likely comes from our app (it's a way of removing the existing answer).

Reply:
> it's a way of removing the existing answer

It doesn't look to me like nulls would remove the existing answer currently.

Reply:
Why? Please look at https://github.com/getodk/collect/pull/4351/files#diff-d4cabe3ff3c1faf046ef950df40de48cbf816cb2ab01e63578d5e8e897bef3e3R50

if there is an answer it is deleted and then below we check if the new answer is file or not and we skip rest of the code.

Reply:
So if `null` is passed in, `answerFile` might still be non-null and in that case it would clear the answer out. Got it. I expected an explicit check on `object == null` but now I see why it's not needed.

Reply:
There is even a test for that case.